### PR TITLE
[Blazor] Adds support for rendering client-side components from MVC

### PR DESCRIPTION
**Pull request by javiercn**
* Adds support for rendering marker comments that blazor.webassembly.js can pick-up in the future to render client-side components.

Addresses https://github.com/aspnet/AspNetCore/issues/14434

### CODE REVIEW

File: src/Mvc/Mvc.ViewFeatures/src/ClientComponentMarker.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+namespace Microsoft.AspNetCore.Components
+{
+    // **Component descriptor protocol**
+    // MVC serializes one or more components as comments in HTML.
+    // Each comment is in the form <!-- Blazor:<<Json>>--> for example { "type": "client", "typeName": "<<type>>", "assembly": "<<assembly>>", "parameterDefinitions": [{"name":"<<name>>", typeName="<<type>>", "assembly": "<<assembly>>"}], "parameterValues": [<<value>>] }
+    // Where <<Json>> has the following properties:
+    // 'type' indicates the marker type. It's value for client-side components is 'client'.
+    // 'typeName' the full type name for the rendered component.
+    // 'assembly' the assembly name for the rendered component.
+    // 'parameterDefinitions' an array that contains the definitions for the parameters including their names and types and assemblies.
```
Comment (rynowak) [inline src/Mvc/Mvc.ViewFeatures/src/ClientComponentMarker.cs]: Why do we need this? The server knows the parameter types already.

File: src/Mvc/Mvc.ViewFeatures/src/ClientComponentMarker.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    // 'parameterValues' an array containing the parameter values.
+    // 'prerenderId' a unique identifier that uniquely identifies the marker to match start and end markers.
+    internal struct ClientComponentMarker
```
Comment (rynowak) [inline src/Mvc/Mvc.ViewFeatures/src/ClientComponentMarker.cs]: How frequently are we going to hit the message size limit in SignalR - have we tested this?

File: src/Mvc/Mvc.ViewFeatures/src/ClientComponentMarker.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Mvc/Mvc.ViewFeatures/src/ClientComponentMarker.cs]: As per other conversation: https://github.com/aspnet/AspNetCore/pull/14465#discussion_r329495458

**Comment by SteveSandersonMS** [comment]
It would be great if we had at least a prototype proof of concept that this is definitely sufficient for booting up the app on the client side. I know it looks like it works in theory, and it most likely actually does, but we'll be unable to fix it if it turns out we needed some different information or we can't deserialize adequately or whatever.

File: src/Mvc/Mvc.ViewFeatures/src/ClientComponentMarker.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Mvc/Mvc.ViewFeatures/src/ClientComponentMarker.cs]: When we come to making this human-writable, example:

```html
<div data-blazor-component="MyApp.MyComponent" data-parameter-MyParam1="Hello" data-parameter-MyOtherThing="{'abc': 123}"></div>
```

... then we will need either:

 * Not to allow passing parameters (which would be a bit sucky)
 * Or for the client app to be able to handle not being told the parameter type in the cases where there are matching `[Parameter]` properties on the component

However that doesn't have to be part of 3.1, as there's no server-code dependency for this.

**Comment by SteveSandersonMS** [comment]
The implementation certainly looks good to me. My main reservation is more about the design concept. I'm still coming to terms with whether it's legit for the server to tell the client which root components to use.

This is only going to work in the case where the server is correctly configured to serve a client-side Blazor app that knows how to render matching component types, but there's nothing to guarantee that is the case. We just hope it is true. More general cases exist though:

 * The server project might not even want to have a runtime reference to the client app assembly. Maybe the client app has dependencies that are harmful to load on the server.
 * There might be many different client-side Blazor apps referenced by the same server app. The new APIs make it *look* as if you can simply reference them all and then make different `Html.RenderComponentAsync` calls in different server-side pages. But to make it really work is far more complicated, because you need to also configure your middleware to serve different client-side apps for different URLs and they all need to match up exactly with your `Html.RenderComponentAsync` calls. I'm concerned people will think we're saying they can do this, find it impractical, and regard it as a bug.
 * The idea that you'd choose to render anything other than something that is, in purpose, a "root" component (containing a `Router` usually) is almost definitely mistaken if you're doing a SPA. People may try to create horrible mash-ups of serverside and clientside routing, whereby you navigate to different server-side pages just to reach different client-side components. I know it's not our job to stop people doing all bad things, but the very idea that the server chooses the client-side root component seems suspect if really there's only one correct client-side root component (which the client app could already know anyway).

These are my reservations, but I'm not trying to veto it. Legitimate scenarios that I'd definitely endorse:

 * Prerendering+client is a valid and desirable combination, and this makes it super easy in the most typical template-like case.
 * People have asked for a way to pass in config-like info from the server to the client app even when not prerendering, and this also gives a very easy way to do that.

**Comment by SteveSandersonMS** [comment]
Actually I've thought of a possibly bigger issue: in .NET 5, with our plan of record for TFMs, it won't even be possible to reference the client-side project from the server-side project.

I'm not sure what this will mean for our hosting APIs, but certainly this is something we should think through before creating a special new server+client startup API that relies on this kind of reference.

@rynowak Since @javiercn is out, let's discuss this. Maybe I'm totally wrong about this, which would be nice.

**Comment by javiercn** [comment]
I'm just pasting the thoughts I have in here to keep a record with the remaining context, but we should discuss this outside of this PR. I'm going to try and see 

If I can convey my points below.

> This is only going to work in the case where the server is correctly configured to serve a client-side Blazor app that knows how to render matching component types, but there's nothing to guarantee that is the case. We just hope it is true. 

This is somewhat true in the server-side case too. You need to setup the script and the hub correctly in your app. When referencing a blazor client-side app, I expect all the metadata to be present so that you can add the script in your page/view. (I'll expand on this later, see appendix A)

> More general cases exist though: 
> * The server project might not even want to have a runtime reference to the client app assembly. Maybe the client app has dependencies that are harmful to load on the server.

This is a fair point, but I'm wondering how common/when this is the case. It is also possible for the server to not even have a runtime reference to the app, but only to the subset it wants to prerender (I'll expand on this later, see appendix B).

> * There might be many different client-side Blazor apps referenced by the same server app. The new APIs make it _look_ as if you can simply reference them all and then make different `Html.RenderComponentAsync` calls in different server-side pages. But to make it really work is far more complicated, because you need to also configure your middleware to serve different client-side apps for different URLs and they all need to match up exactly with your `Html.RenderComponentAsync` calls. I'm concerned people will think we're saying they can do this, find it impractical, and regard it as a bug.

* I think having multiple client rendered apps is a more advanced scenario and won't be that common, but even so, it should be a matter of including the scripts on the page to load the right app.
* My expectation is that blazor client-side apps are served as plain static files and that we don't need any runtime support (other than SWA) to make them work.

> * The idea that you'd choose to render anything other than something that is, in purpose, a "root" component (containing a `Router` usually) is almost definitely mistaken if you're doing a SPA. People may try to create horrible mash-ups of serverside and clientside routing, whereby you navigate to different server-side pages just to reach different client-side components. I know it's not our job to stop people doing all bad things, but the very idea that the server chooses the client-side root component seems suspect if really there's only one correct client-side root component (which the client app could already know anyway).

I think this is a discussion about scenarios. So I'll detail my views on scenarios (with examples) and how I envision them working:
* **Prerendering a client-side app:** Let's imagine we have a client-side app hosted inside an ASP.NET Core app that offers an API that the client-side app uses. I would expect this to be a common scenario (its one of our starting points after all)
  * We will have two/three projects:
    * Client for rendering client-side
    * Server the host for the API and for serving the client-side app.
    * Shared for shared logic/DTOs
  * The client app is just static files that are being served by the Server app.
    * We just reference the client-side app from the server-side app and bring all the assets from the client app.
  * I think it will be desirable to allow prerendering to work without having to change the project structure.
    * To that effect, we can leverage the existing reference and "Prerender" the root component.
      * This will work well as a starting point. Once the app grows there will inevitably be some work on the user to optimize this experience, but the work and the pattern to follow are well defined and is a prevalent topic on all SPA frameworks.
        * Hide any platform specific code that needs to run in both places under an abstraction that can be replaced on the server. I don't think this will be common.
          * For example if you use local storage or similar, it needs to be replaceable. Most of these things will only be used inside OnAfterRender as its the only place where we guarantee interop works, and that doesn't get invoked during prerendering.
        * Optimize some workloads for prerendering: In some cases you will want to skip issuing an HTTP request to your API when prerendring and go straight to the DB or similar. I don't think this will be common as it requires extra work and complicates things in the long run.
* **Rendering interactive widgets from an MVC/Razor pages app:** In this scenario we have a traditional MVC/Razor pages app that wants to be enhanced with some interactive components. Lets assume that this app wants to use the component suite from a third-party vendor.
  * We will have two projects:
    * Client for rendering client-side
    * Server the host for the API and for serving the client-side app.
  * Both client and server will have a reference to the component library. When a page is being rendered, the server will include one or more components into the page.
  * There is no need for the server to reference the client app in this scenario, as the components are in the shared library.
* **Transitioning an app from MVC to Blazor:** In this scenario we have an existing MVC/Razor pages app that we want to transition to blazor client-side.
* Initially we have one project (the MVC app) and we create two new projects, a client-side app and a library for shared code.
* We start replacing MVC primitives with Components primitives on a per page basis.
  * Routing might still happen server-side, but forms and things like that get handled client-side and switched to invoke APIs on the server instead of form posts.
  * When a page is being rendered it decides what components it needs to include and references the client-side app assets to bootstrap it.

## Appendix A
* My expectation when running blazor apps is that they are just static files served from wwwroot and that we don't need any additional runtime support for serving published blazor apps in production.
* I don't expect to have multiple Blazor apps running on the same rendered document/page given that there is a limitation of 1 CLR instance per wasm, although this might be lifted in the future.
* I expect rendering a client-side app to simply require including the blazor wasm scripts and some metadata that points to a manifest that the script uses to boot the app. Something like `<script src="blazor.wasm.js" data-manifest="clientappname.manifest.json"></script>`.
* I expect that normally we have a single client-side app that is served from a server app, whether it is a full-SPA app, a traditional app that is transitioning to become a SPA or a traditional app that render some interactive Widgets.
* The way I expect this to work is for the server app to render the client-side app and for both apps to have a common subset of referenced libraries (for example third party vendor component libraries)

## Appendix B
* The app can always be structured in a way where there is a common library that holds a subset of the functionality that is shared between the two projects and that doesn't have any harmful client specific dependency.
  * For example, you can move all the components to a class library and have the client-side app be just a shell for the app, with all/most of the code living in the shared class library. This is how Xamaring and UWP tackle the problem AFAIK.
* WRT to avoiding a reference to the client app from the server, I expect this to be a more advanced scenario but it depends on where we land on TFMs/RIDs.
  * It would force us to do something weird in the project template:
    * We need some piece of data in the CSPROJ to relate the client and server projects for build, dependency ordering, publish, etc.
    * We would have to create our own custom reference (BlazorReference) which is a lot of work and forces a new concept into users or we would have to do something custom on the ProjectReference item, like ReferenceOutputAssembly=false.
  * We can't support prerendering without running the code on the server, and I think we would like that to be either enabled by default or trivially achievable.

**Comment by javiercn** [comment]
Closing as I didn't get time to develop the idea further for 3.1. We'll revisit alternatives for blazor-wasm

**Pull request closed without merging**